plugins/woocommerce/src/Internal/Admin/BlockTemplateRegistry/BlockTemplateRegistry.php

```diff
@@ -32,36 +34,45 @@ public static function get_instance(): BlockTemplateRegistry {
 		return self::$instance;
 	}
 
-    /**
-     * Register a single template.
-     *
-     * @param string $id Template ID.
-     * @param array  $template Template layout.
-     */
-    public function register( BlockTemplateInterface $template ) {
-        $id = $template->get_id();
+	/**
+	 * Register a single template.
+	 *
+	 * @param BlockTemplateInterface $template Template to register.
+	 *
+	 * @throws \ValueError If a template with the same ID already exists.
+	 */
+	public function register( BlockTemplateInterface $template ) {
+		$id = $template->get_id();
 
-        if ( isset( $this->templates[ $id ] ) ) {
+		if ( isset( $this->templates[ $id ] ) ) {
 			throw new \ValueError( 'A template with the specified ID already exists in the registry.' );
 		}
 
-        $this->templates[ $id ] = $template;
-    }
+		/**
+		 * Fires when a template is registered.
+		 *
+		 * @param BlockTemplateInterface $template Template that was registered.
+		 *
+		 * @since 8.2.0
+		 */
+		do_action( 'woocommerce_block_template_register', $template );
```

Comment:
Nice!  I like this action overall, though I do have a question/concern about the example usage in this PR.

Using this filter as a hook to add blocks to existing groups/sections implies that those blocks are registered prior to the template being registered.

While the hook itself is convenient, this only works if those blocks have been added first.  My concern is that 3PDs may rely on this assuming that all blocks are registered, which may not be the case.

Reply:
Is this accurate? I thought that the registering of the template could be done at any time and it was only by the point that the template was used on the front end that the blocks had to actually be registered?

The BlockTemplateRegistry and the template API don't require registration of blocks.

All that said, and assuming I am not incorrect... is there an existing hook that you think would be better to use?

Reply:
>The BlockTemplateRegistry and the template API don't require registration of blocks.

This is the exact reason I don't think this pattern works very well for registration of third party blocks.  In your example above we have:

```php
$basic_details = $template->get_section_by_id( 'basic-details' );
```

There is no guarantee that the `basic-details` section has been added when the template is registered since the section could be added easily after registration.  We can make a deliberate choice to register those fields prior to registering the template, but these leaves some room for developer error.

This is also an issue for 3PDs who want to register a field within another plugin's third party section since they will have to be aware of which hook and priority are being used in order to register their block after that section has been added.  This type of "race condition" will exist with most hooks, but may be a less common case given it occurs with third party plugins extending other third party plugins.

If our goal is a guarantee that a parent block does exist prior to our block additions, one possible option would be a hook that occurs on block registration.

```php
apply_filters( 'woocommerce_template_block_registered', $block );

add_filter( 'woocommerce_template_block_registered', 'add_my_block' );
function add_my_block( $block ) {
    if ( $block->get_id() !== 'target-parent' ||$block->get_template() !== 'target-template' ) {
        return;
    }
    $block->add_block( ... );
}
```

Reply:
> There is no guarantee that the basic-details section has been added when the template is registered since the section could be added easily after registration. We can make a deliberate choice to register those fields prior to registering the template, but these leaves some room for developer error.

Currently the way we are approaching things, we create the entire template and then register it. Do you have any examples of scenarios where we would not want to do that (and the priority approach wouldn't work)?

I agree that if we were to then modify the template that we (and third-parties) would have to be aware of priorities (and what hook, if we "support" templates being modified from any hook... if we strongly encourage template modification to always be done in the `woocommerce_block_template_register` hook, then they "just" need to worry about the priorities for that hook).

> If our goal is a guarantee that a parent block does exist prior to our block additions, one possible option would be a hook that occurs on block registration.

I'm not sure I understand what you mean by "registered" in the context of an individual block (instance)... do you mean whenever a block instance is added to a template (via `add_block`, etc.)? So, whenever a block was added to a template, we would fire `woocommerce_template_block_registered`?

If we did this, we would also have to have a similar hook that fired when a block was removed, `woocommerce_template_block_unregistered`.

I think that things will quickly get confusing with this approach, and while if properly coded, extensions could "guarantee" that a parent existed before they added their dependent blocks, I question whether it is worth the extra complication over a more simple approach where they just worry about what priority their `woocommerce_block_template_registered` hook is called.

And, I wonder if there would be scenarios where an extension wanted to make a decision whether to make template modifications based on more than one existing block... if that were the case, the extension would need to keep track of which blocks had been registered and only do their modification after they had seen all those blocks added (and, as mentioned above they would also have to be concerned about those blocks being removed).

It's an interesting dilemma.

For third-party blocks that do *not* depend on the existence of any other blocks, would you see them using the `woocommerce_block_template_registered` hook like this PR introduces? If not, what would you see them using?

Reply:
>Currently the way we are approaching things, we create the entire template and then register it. Do you have any examples of scenarios where we would not want to do that (and the priority approach wouldn't work)?

I guess my issue with this is that it relies heavily on whoever is adding the template to understand that blocks should all be added prior to registration and nothing about the API implies that.

We don't have a specific case right now, but there's also nothing stopping anyone from registering a template first and adding blocks later.  We also know for a fact that third party blocks will have to be added after the template has been registered.

>I'm not sure I understand what you mean by "registered" in the context of an individual block (instance)... do you mean whenever a block instance is added to a template (via add_block, etc.)? So, whenever a block was added to a template, we would fire woocommerce_template_block_registered?

Correct.  Poor choice of semantics on my part.  `woocommerce_template_block_added` would be more appropriate to distinguish between block (type) registration and addition of a block to a template.

>I think that things will quickly get confusing with this approach

Could you expand on what you feel might be confusing with this approach?

>For third-party blocks that do not depend on the existence of any other blocks, would you see them using the woocommerce_block_template_registered hook like this PR introduces?

I think that's okay for most cases, but we there are still 2 issues that stand out in my mind:
* That third party does not have any way of knowing if other third party plugins might depend on this block
* We would need some better way of enforcing/encouraging all core blocks to be added prior to template registration

Reply:
> Could you expand on what you feel might be confusing with this approach?

I think it gets confusing and complicated when block removal and any dependency on multiple blocks is involved, as I had mentioned above.

> That third party does not have any way of knowing if other third party plugins might depend on this block

What would one do (in code) with the knowledge that another third party depended on their block?

Reply:
>I think it gets confusing and complicated when block removal and any dependency on multiple blocks is involved, as I had mentioned above.

Could you give a specific case where this becomes more confusing?  To me it seems to simplify things greatly, but I may be missing something you're seeing.

>What would one do (in code) with the knowledge that another third party depended on their block?

This is just referring to the issue where plugins need to be highly aware of timing (priority in this case) and a plugin should probably avoid changing their priority since this could affect plugins downstream that are relying on that plugin's action firing at a specific priority.

In this case I think that if we think of hooks as events (which they are), it can help us to identify that plugins wanting to add their blocks to other blocks are more concerned with the event of block addition over the event of template registration.  But I'm curious to see more specific examples where this might be confusing so we can identify how to clean that up.